File: lib/extractor.ts
import * as cheerio from "cheerio";
import { AnyNode, Element, Text } from "domhandler";

export interface Match {
  match_id: string;
  team_1: string[];
  team_2: string[];
  score_1: number;
  score_2: number;
  winner_team: 0 | 1 | 2;
}

// ── Fetch ─────────────────────────────────────────────────────────────

async function fetchHtml(url: string): Promise<string> {
  const res = await fetch(url, {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) " +
        "AppleWebKit/537.36 (KHTML, like Gecko) " +
        "Chrome/126.0.0.0 Safari/537.36",
    },
  });
  if (!res.ok) throw new Error(`Fetch failed: ${res.status} ${res.statusText}`);
  return res.text();
}

// ── Public API ────────────────────────────────────────────────────────

export async function parseSession(url: string): Promise<Match[]> {
  const html = await fetchHtml(url);
  return parseMatches(html);
}

export function parseMatches(html: string): Match[] {
  const $ = cheerio.load(html);
  const matches = tryStructuredParse($);
  if (matches.length > 0) return matches;
  return flatTextParse($);
}

// ── Structured DOM parse ──────────────────────────────────────────────

function tryStructuredParse($: cheerio.CheerioAPI): Match[] {
  const matches: Match[] = [];

  // Find all text nodes containing "Round"
  $("*").contents().each((_, node) => {
    if (node.type !== "text") return;
    if (!/^\s*Round\s*$/i.test(node.data ?? "")) return;

    const card = findCardAncestor($, node as AnyNode);
    if (!card) return;

    const tokens = getStrippedStrings($, card);
    const match = extractMatchFromTokens(tokens);
    if (match) {
      const round = match.round;
      delete (match as any).round;
      matches.push({ ...match, match_id: `R${round ?? "?"}` } as Match);
    }
  });

  return matches;
}

function findCardAncestor(
  $: cheerio.CheerioAPI,
  node: AnyNode,
  maxDepth = 5
): Element | null {
  let current = node.parent as Element | null;

  for (let i = 0; i < maxDepth; i++) {
    if (!current || (current.type !== "tag" && current.type !== "script" && current.type !== "style")) return null;
    if (current.name === "body") return null;

    const texts = getStrippedStrings($, current);
    const digitCount = texts.filter((t) => /^\d+$/.test(t)).length;
    if (digitCount >= 2 && texts.some((t) => t.includes("Round"))) {
      return current;
    }
    current = current.parent as Element | null;
  }

  return null;
}

function getStrippedStrings($: cheerio.CheerioAPI, el: AnyNode): string[] {
  const results: string[] = [];
  $(el)
    .contents()
    .each(function collectText(_: number, node: AnyNode) {
      if (node.type === "text") {
        const text = (node as Text).data?.trim();
        if (text) results.push(text);
      } else if (node.type === "tag" || node.type === "script" || node.type === "style") {
        $(node)
          .contents()
          .each(function (this: AnyNode, i: number, child: AnyNode) {
            collectText.call(this, i, child);
          });
      }
    });
  return results;
}

// ── Token parsing (shared between structured + flat) ──────────────────

interface RawMatch {
  round?: number;
  team_1: string[];
  team_2: string[];
  score_1: number;
  score_2: number;
  winner_team: 0 | 1 | 2;
}

function extractMatchFromTokens(tokens: string[]): RawMatch | null {
  // Merge "Court" + digit tokens
  const merged: string[] = [];
  let i = 0;
  while (i < tokens.length) {
    if (
      tokens[i].toLowerCase() === "court" &&
      i + 1 < tokens.length &&
      /^\d+$/.test(tokens[i + 1])
    ) {
      merged.push(`Court ${tokens[i + 1]}`);
      i += 2;
    } else {
      merged.push(tokens[i]);
      i++;
    }
  }
  tokens = merged;

  let roundNum: number | undefined;
  let startIdx = 0;

  for (let j = 0; j < tokens.length; j++) {
    if (
      tokens[j].toLowerCase() === "round" &&
      j + 1 < tokens.length &&
      /^\d+$/.test(tokens[j + 1])
    ) {
      roundNum = parseInt(tokens[j + 1], 10);
      startIdx = j + 2;
      break;
    }
  }

  if (startIdx < tokens.length && tokens[startIdx].toLowerCase().startsWith("court")) {
    startIdx++;
  }

  const remaining = tokens.slice(startIdx);
  const segments: [string[], number][] = [];
  let currentNames: string[] = [];

  for (const t of remaining) {
    if (/^\d+$/.test(t)) {
      if (currentNames.length > 0) {
        segments.push([currentNames, parseInt(t, 10)]);
        currentNames = [];
      }
    } else {
      currentNames.push(t);
    }
  }

  if (segments.length < 2) return null;

  const [team1Players, score1] = segments[0];
  const [team2Players, score2] = segments[1];

  if (!team1Players.length || !team2Players.length) return null;

  return {
    round: roundNum,
    team_1: team1Players,
    team_2: team2Players,
    score_1: score1,
    score_2: score2,
    winner_team: score1 > score2 ? 1 : score2 > score1 ? 2 : 0,
  };
}

// ── Flat-text fallback parse ──────────────────────────────────────────

function flatTextParse($: cheerio.CheerioAPI): Match[] {
  const allText: string[] = [];
  $("*")
    .contents()
    .each((_, node) => {
      if (node.type === "text") {
        const t = (node as Text).data?.trim();
        if (t) allText.push(t);
      }
    });

  const matches: Match[] = [];
  let i = 0;

  while (i < allText.length) {
    if (allText[i].toLowerCase() === "round") {
      const window = allText.slice(i, i + 12);
      const match = extractMatchFromTokens(window);
      if (match) {
        const round = match.round;
        delete (match as any).round;
        matches.push({ ...match, match_id: `R${round ?? "?"}` } as Match);
        i += 8;
        continue;
      }
    }
    i++;
  }

  return matches;
}

File: _legacy/extractor.py
# extractor.py
"""
Reclub scoresheet parser — fetches and parses doubles match results.
Designed as a module: import parse_session() or run standalone for testing.
"""

import re
import sys
from collections import defaultdict

import requests
from bs4 import BeautifulSoup, Tag


def fetch_html(url: str) -> str:
    headers = {
        "User-Agent": (
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
            "AppleWebKit/537.36 (KHTML, like Gecko) "
            "Chrome/126.0.0.0 Safari/537.36"
        )
    }
    resp = requests.get(url, headers=headers, timeout=15)
    resp.raise_for_status()
    return resp.text


def parse_matches(html: str) -> list[dict]:
    soup = BeautifulSoup(html, "html.parser")
    matches = _try_structured_parse(soup)
    if matches:
        return matches
    return _flat_text_parse(soup)


def parse_session(url: str) -> list[dict]:
    """Convenience: fetch + parse in one call."""
    return parse_matches(fetch_html(url))


# ── Structured DOM parse ─────────────────────────────────────────────

def _try_structured_parse(soup: BeautifulSoup) -> list[dict]:
    matches = []
    round_markers = soup.find_all(string=re.compile(r"^\s*Round\s*$", re.I))

    for marker in round_markers:
        card = _find_card_ancestor(marker)
        if card is None:
            continue
        tokens = [t.strip() for t in card.stripped_strings if t.strip()]
        match = _extract_match_from_tokens(tokens)
        if match:
            match["match_id"] = f"R{match.pop('round', '?')}"
            matches.append(match)
    return matches


def _find_card_ancestor(node, max_depth: int = 5) -> Tag | None:
    current = node.parent
    for _ in range(max_depth):
        if current is None or current.name == "body":
            return None
        texts = list(current.stripped_strings)
        digit_count = sum(1 for t in texts if re.fullmatch(r"\d+", t))
        if digit_count >= 2 and any("Round" in t for t in texts):
            return current
        current = current.parent
    return None


def _extract_match_from_tokens(tokens: list[str]) -> dict | None:
    merged = []
    i = 0
    while i < len(tokens):
        if tokens[i].lower() == "court" and i + 1 < len(tokens) and tokens[i + 1].isdigit():
            merged.append(f"Court {tokens[i + 1]}")
            i += 2
        else:
            merged.append(tokens[i])
            i += 1
    tokens = merged

    round_num = None
    start_idx = 0
    for i, t in enumerate(tokens):
        if t.lower() == "round" and i + 1 < len(tokens) and tokens[i + 1].isdigit():
            round_num = int(tokens[i + 1])
            start_idx = i + 2
            break

    if start_idx < len(tokens) and tokens[start_idx].lower().startswith("court"):
        start_idx += 1

    remaining = tokens[start_idx:]
    segments = []
    current_names = []
    for t in remaining:
        if re.fullmatch(r"\d+", t):
            if current_names:
                segments.append((current_names, int(t)))
                current_names = []
        else:
            current_names.append(t)

    if len(segments) < 2:
        return None

    team1_players, score1 = segments[0]
    team2_players, score2 = segments[1]
    if not team1_players or not team2_players:
        return None

    return {
        "round": round_num,
        "team_1": team1_players,
        "team_2": team2_players,
        "score_1": score1,
        "score_2": score2,
        "winner_team": 1 if score1 > score2 else (2 if score2 > score1 else 0),
    }


# ── Flat-text fallback parse ─────────────────────────────────────────

def _flat_text_parse(soup: BeautifulSoup) -> list[dict]:
    all_text = list(soup.stripped_strings)
    matches = []
    i = 0
    while i < len(all_text):
        if all_text[i].lower() == "round":
            window = all_text[i : i + 12]
            match = _extract_match_from_tokens(window)
            if match:
                match["match_id"] = f"R{match.pop('round', '?')}"
                matches.append(match)
                i += 8
                continue
        i += 1
    return matches


# ── Standalone test ──────────────────────────────────────────────────

if __name__ == "__main__":
    url = sys.argv[1] if len(sys.argv) > 1 else "https://api.reclub.co/export/score-sheets?m=7BNX41"
    print(f"Fetching: {url}\n")
    matches = parse_session(url)

    if not matches:
        print("⚠  No matches found.")
    else:
        print(f"Parsed {len(matches)} matches:\n")
        for m in matches:
            t1 = " & ".join(m["team_1"])
            t2 = " & ".join(m["team_2"])
            w = "←" if m["winner_team"] == 1 else ("→" if m["winner_team"] == 2 else "=")
            print(f"  {m['match_id']:>4s}  {t1:<25s} {m['score_1']} {w} {m['score_2']}  {t2}")
